intake_esm/core.py

```diff
@@ -145,98 +94,194 @@ def unique(self, columns=None):
         return info
 
     def __repr__(self):
-        """Making string representation of object."""
+        """Make string representation of object."""
         info = self.nunique().to_dict()
         output = []
         for key, values in info.items():
             output.append(f'{values} {key}(s)\n')
         output = '\n\t> '.join(output)
-        items = len(self.ds.index)
-        return f'{self.collection_name.upper()} collection catalogue with {items} entries:\n\t> {output}'
+        items = len(self.df.index)
+        return f'ESM Collection with {items} entries:\n\t> {output}'
+
+
+class ESMDatasetSource(intake_xarray.base.DataSourceMixin):
+    container = 'xarray'
+    name = 'esm-dataset-source'
+
+    def __init__(self, esmcol_path, query, **kwargs):
+        super().__init__(metadata={})
+        self.esmcol_path = esmcol_path
+        self._col_data = _fetch_and_parse_file(esmcol_path)
+        self.df = self._get_subset(**query)
+        self.urlpath = ''
+        self._ds = None
+
+    def _get_subset(self, **query):
+        df = pd.read_csv(self._col_data['catalog_file'])
+        if not query:
+            return pd.DataFrame(columns=df.columns)
+        condition = np.ones(len(df), dtype=bool)
+        for key, val in query.items():
+            if isinstance(val, list):
+                condition_i = np.zeros(len(df), dtype=bool)
+                for val_i in val:
+                    condition_i = condition_i | (df[key] == val_i)
+                condition = condition & condition_i
+            elif val is not None:
+                condition = condition & (df[key] == val)
+        query_results = df.loc[condition]
+        return query_results
+
+    def to_xarray(self, zarr_kwargs={}, cdf_kwargs={}):
```

Comment:
I'm a bit concerned about the name of this method. `to_xarray` returns a dict type. I think this will lead to confusion (just as `to_dask` returning a xr.Dataset has). Also, pandas has a `to_xarray` method and it returns a xarray.DataArray or xarray.Dataset.

```python
> import xarray as xr
> dsets = cat.to_dask()
> print(type(cat))
> print(type(dsets))
> print(isinstance(dsets, (xr.DataArray, xr.Dataset)))

<class 'intake.catalog.local.LocalCatalogEntry'>
<class 'dict'>
False
```

Would `to_datasets` be a better name?

Reply:
I like `to_datasets`.